students/jeff_shabani/session07/html_render.py

```diff
@@ -3,26 +3,204 @@
+            file_name.write(f' {key}="{value}"')
+        file_name.write(f'{self._front_tag()[-1:]}\n')
+
+        for content in self.content:
+            if hasattr(content, 'render'):
+                content.render(file_name, indent, ind_count + 1)
+            else:
+                file_name.write(f'{self.indent}{indent}{content}\n')
+
+        file_name.write(f'{self.indent}{self._end_tag()}\n')
+
+
+class Html(Element):
+    """
+    html sublcass
+    """
+    tag = 'html'
+
+    def render(self, file_name, indent="", ind_count=0):
+        self.indent = indent
+        file_name.write(f'<!DOCTYPE html>\n')
+        super().render(file_name, indent)
+
+
+class P(Element):
+    """
+    class for p tag
+    """
+    tag = 'p'
+
+
+class Body(Element):
+    """
+    class for p tag
+    """
+    tag = 'body'
+
+
+"""
+Step 3: print on one line
+"""
+
+
+class OneLineTag(Element):
 
     def append(self, new_content):
-        result = list()
-        result.append(new_content)
-        for i in result:
-            self.content +=i
-        return self.content
 
+        raise NotImplementedError
+
+    def render(self, file_name, indent="", ind_count=0):
+
+        self.indent = indent * ind_count
+
+        file_name.write(f'{self.indent}{self._front_tag()[:-1]}')
+        for key, value in self.attrs.items():
+            file_name.write(f' {key}="{value}"')
+        file_name.write(f'{self._front_tag()[-1:]}')
+        file_name.write(f'{self.content[0]}')
+        file_name.write(f'{self._end_tag()}\n')
+
+
+class Head(Element):
+    tag = "head"
+
+
+"""
+Step 5: Self closing tag
+"""
+
+
+class SelfClosingTag(Element):
+
+    def __init__(self, content=None, **kwargs):
+
+        if content: raise TypeError
+        self.attrs = kwargs
+
+    def append(self, new_content):
+        raise NotImplementedError
+
+
+    def render(self, file_name, indent="", ind_count=0):
+
+        """
+        if content is entered this tells user that self closing tags
+        can't have conent and resets the conent to an empty string.
+        """
+
+        self.indent = indent * ind_count
+
+        file_name.write(f'{self.indent}{self._front_tag()[:-1]}')
+        for k, v in self.attrs.items():
+            file_name.write(f' {k}="{v}"')
+        file_name.write(f' />\n')
+
+
+"""
+Step 6
+"""
+
+
+class A(OneLineTag):
+    tag = 'a'
+
+    def __init__(self, link, content=None, **kwargs):
+        if not (content and link):
+            raise TypeError
+
+        kwargs['href'] = link
+        super().__init__(content, **kwargs)
+
+
+class Hr(SelfClosingTag):
+
+    tag = 'hr'
+
+
+# <br /> is XHTML format
+class Br(SelfClosingTag):
+    tag = 'br'
+
+    def __init__(self, content=None, **kwargs):
+
+        if kwargs: raise TypeError
+
+        super().__init__(content, **kwargs)
+
+
+class Ul(Element):
+    tag = 'ul'
+
+    def __init__(self, content=None, **kwargs):
+
+        if content: raise TypeError
+
+        self.contents = []
```

Comment:
Hi Jeff,

Everything is working correctly, except for this line. Everywhere else in your code, you're using _self.content_ (singular), while this one uses _self.contents_ (plural). Your inherited append method does this:

```
    def append(self, new_content):
        self.content.append(new_content)
```

Which means you will trigger an attribute error when trying to do append on a _Ul_ object. It's a tiny change, so I'm guessing for some reason this one did not make it into your final commit. Once that error is fixed, your render works correctly all the way to the end. No need to create an additional pull request, just make the correction and add it to future pull request, as part of another assignment.

-Luis